design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,327 @@
+depend on the release stage (Alpha, Beta, GA).
+
+[Removal](#removal) of features is widely discussed later
+in this proposal.
+
+#### Enhancement proposal
+As the first step for introducing a new feature, a formal proposal is
+expected to be shared for public review via mailinglist and
+a [design proposal](https://github.com/kubevirt/community/tree/main/design-proposals).
+
+This is the first opportunity to evaluate a new feature.
+The proposal needs to include motivation, goals, implementation details
+and phases. Review the [proposal template](https://github.com/kubevirt/community/blob/main/design-proposals/proposal-template.md)
+for more information.
+
+#### Implementation
+The development work on the feature is expected to include coding,
+testing, integration and documentation.
+
+#### Releases
+- **Alpha**:
+  An initial release of the feature for experimental purposes.
+  Recommended for non-production usages, evaluation or testing.
+
+  The API is considered unstable and may change significantly.
+  There are no backward compatability considerations and it can
+  be removed at any time.
+
+  The period in which a feature can remain in Alpha is limited,
+  assuring features are not piling up without control.
+  See [release stage transition table](#release-stage-transition-table)
+  for more information.
+
+  The feature presence is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **Beta**:
+  The first release that can be evaluated with care in production.
+  Acting as a pre-release, its main objective is to collect feedback
+  from users to assure its usefulness and readiness for graduation.
+  If there is no confidence of usage or usefulness, it may remain in
+  this stage for some time.
+
+  However, the period in which a feature can remain in Beta is limited,
+  assuring features are not piling up without control.
+  See [release stage transition table](#release-stage-transition-table)
+  for more information.
+
+  The API is considered stable with care not to break backward compatibility
+  with previous beta releases.
+  This implies that fields may only be added during this stage,
+  not removed or renamed.
+
+  The feature presence is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **GA**:
+  The feature graduated to general-availability (GA) and is now part of
+  the core features.
+
+  The API is considered stable with care not to break backward compatibility
+  with the previous releases.
+
+  The feature functionality is no longer controlled by a FG.
+
+> **Warning**: A Feature Gate flag is solely intended to control
+> feature lifecycle. It should not be confused and used as a cluster
+> configurable enablement of the functionality.
+> In cases where the cluster admin should control a functionality,
+> regardless to the feature stage, dedicated configuration field/s
+> should be included.
```

Comment:
I am not really sure why we would need additional mechanism here, Can you elaborate?

Reply:
It is described at `### Terminology`.

If the need is to control a feature settings, including its activation/deactivation at the operational level and not as part of its lifecycle, then it must be provided through a different mechanism.

Considering that a FG is eventually removed, either because the feature got dropped or graduated, the Admin may still want to control it.
It is surely not required to have such an option and indeed many features do not have it.

Reply:
Anything else that is needed to resolve this conversation=

Reply:
I am not sure we need another field for this and why can't the feature gate stay around if it is graduated but deemed to be not for all use cases.

Reply:
Perhaps it will help if you can feedback on which part of the definition of "what a FG is" you disagree or it is not clear.

> ### Terminology
> Feature Gates and feature configuration are often used interchangeable or not differentiated.
> - Feature Gate: A flag that controls the presence or availability of a feature in the cluster.
> - Feature configuration: Cluster or workload level configuration that allows an admin or user
>   (depending on the feature) to control aspects of a feature operation.
>   A common usage is to determine if features are opt-in or opt-out by default.

If you are looking for an U/S ref, then see the Kuberenetes FG stages [1].

It states:

> A General Availability (GA) feature is also referred to as a stable feature. It means:
> 
>  - **The feature is always enabled; you cannot disable it.**
>  - **The corresponding feature gate is no longer needed.**
>  - Stable versions of features will appear in released software for many subsequent versions

Once GA (or removed) the FG is no longer controlling the feature.
Which implies you need a different operational knob to do so, if it is desired.

[1] https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages

Reply:
So what I fear is that we are going to have a tons of new fields for each feature that is GAed just so we can turn it off/on. I am not sure if I understand this correctly or we aim to have a general mechanism that can be reused.

Reply:
The feature owners should decide if such a knob is needed or not.
It makes sense that during the design review period the proposal will be convincing with reasoning on why it is needed.

Any API addition, including such knobs should be minimized. But this is not directly related to this proposal.

Reply:
>The feature owners should decide if such a knob is needed or not.

But should we have field per each knob or can a existing or new generic field/mechanism be used?  

I have not look into how K8S does this, would you know of example?

Reply:
Resolved